### PR TITLE
Fix #5150: Correct home screen topic grid misalignment after returning from lesson

### LINKED ISSUE #5150: [BUG]: Home screen topics Grid view unaligned after coming back from Lesson
### Describe the bug

Home screen topics Grid view unaligned after coming back from Lesson

https://github.com/oppia/oppia-android/assets/76530270/48d916d5-6ed1-4914-9d16-70c865d83c0f

### Steps To Reproduce

Go to any topic and then any lesson
Go back by (System or Navigation)
See the Topic Grids now

### Expected Behavior

Topic Grid should remain aligned.

### Screenshots/Videos

_No response_

### What device/emulator are you using?

Poco M2 Pro

### Which Android version is your device/emulator running?

Android 11

### Which version of the Oppia Android app are you using?

0.11-oppia-f0837d22fe

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by TanishMoral11**
## Explanation
Fixes #5150 , where the home screen topic grid becomes misaligned after returning from a lesson. The grid alignment is now maintained upon returning to the home screen, ensuring a consistent layout for topics.

## Issue Description
The home screen topics grid view experiences misalignment when navigating back from a lesson. 
This issue stems from inconsistent layout parameter management within the margin binding adapters, causing unexpected UI disruptions across different screen configurations.

## Changes Made
- Refactored `MarginBindingAdapters` to prioritize `setLayoutParams()` over `requestLayout()`
- Implemented a more robust approach to setting and preserving layout parameters
- Ensured consistent margin application across various screen orientations and device types

## Reasoning
The core of the fix lies in how layout parameters are managed:
- `View.setLayoutParams()` explicitly sets layout parameters for a view
- This method inherently calls `requestLayout()`, guaranteeing the view is redrawn with updated parameters
- Direct parameter setting provides more predictable margin management across different screen configurations

## Test Coverage
Comprehensive testing was added to `MarginBindingAdaptersTest` to validate:
- Margin preservation across multiple screen configurations
- Accurate application of start, end, top, and bottom margins
- Consistency of layout parameter management
- Robust handling of different device orientations (portrait, landscape, tablet)

## Essential Checklist
- [x] The PR title and explanation each start with "Fix #5150: ".
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

### CODE REVIEW

**Comment by TanishMoral11** [comment]
@adhiamboperes PTAL

**Comment by adhiamboperes** [comment]
Thanks @TanishMoral11. Could you PTAL at https://github.com/oppia/oppia-android/issues/5150#issuecomment-1794548399

**Comment by TanishMoral11** [comment]
Hey @adhiamboperes, 
Thank you for your feedback!, I’m a bit unsure about the exact process for testing the HomeActivity as you mentioned. Could you kindly guide me on how to set up the test for this issue? Any specific steps or tips on replicating the behavior and analyzing the layout in Robolectric would be greatly appreciated!

Thanks!

**Comment by TanishMoral11** [comment]
Hey @adhiamboperes ,I’ve written the required tests. Could You PTAL !!
All Test Passing :
image

File: app/src/main/java/org/oppia/android/app/databinding/MarginBindingAdapters.java [hunk 1]
```diff
@@ -6,63 +6,56 @@
...
+  /**
+   * Sets the end margin for a view, accounting for RTL scenarios.
+   */
   @BindingAdapter("layoutMarginEnd")
   public static void setLayoutMarginEnd(@NonNull View view, float marginEnd) {
     if (view.getLayoutParams() instanceof MarginLayoutParams) {
       MarginLayoutParams params = (MarginLayoutParams) view.getLayoutParams();
       MarginLayoutParamsCompat.setMarginEnd(params, (int) marginEnd);
-      view.requestLayout();
+      view.setLayoutParams(params);
```
Comment (adhiamboperes) [inline app/src/main/java/org/oppia/android/app/databinding/MarginBindingAdapters.java:28]: It is worth adding an explanation to the PR descripion describing why and how this change fixes the issue.

File: app/src/main/java/org/oppia/android/app/databinding/MarginBindingAdapters.java [hunk 2]
```diff
@@ -6,63 +6,56 @@
...
       params.topMargin = (int) marginTop;
       view.setLayoutParams(params);
-      view.requestLayout();
     }
   }
 
-  /** Used to set a margin-bottom for views. */
+  /**
+   * Used to set a margin-bottom for views.
+   */
```
Comment (adhiamboperes) [inline app/src/main/java/org/oppia/android/app/databinding/MarginBindingAdapters.java]: All the KDocs in this file seem to be able to fit as single-line Kdocs, as they originally were. They did not need to be changed.

File: app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindingAdaptersTest.kt [hunk 3]
```diff
@@ -292,6 +296,78 @@ class MarginBindingAdaptersTest {
     assertThat(textView.marginEnd.toFloat()).isWithin(TOLERANCE).of(40f)
   }
 
+  @Config(qualifiers = "port")
+  @Test
+  fun testMarginBindableAdapters_layoutParamsSettingPreservesMargins() {
```
Comment (adhiamboperes) [inline app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindingAdaptersTest.kt]:
```suggestion
  fun testMarginBindableAdapters_setLayoutParams_preservesMargins() {
```

File: app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindingAdaptersTest.kt [hunk 4]
```diff
@@ -292,6 +296,78 @@ class MarginBindingAdaptersTest {
...
+    assertThat(textView.marginEnd.toFloat()).isWithin(TOLERANCE).of(40f)
+
+    val layoutParams = textView.layoutParams as ViewGroup.MarginLayoutParams
+    assertThat(layoutParams.topMargin.toFloat()).isWithin(TOLERANCE).of(16f)
+    assertThat(layoutParams.bottomMargin.toFloat()).isWithin(TOLERANCE).of(32f)
+  }
+
+  @Config(qualifiers = "land")
+  @Test
+  fun testMarginBindableAdapters_layoutParamsSettingPreservesMargins_landscape() {
```
Comment (adhiamboperes) [inline app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindingAdaptersTest.kt]:
```suggestion
  fun testMarginBindableAdapters_landscapeMode_setLayoutParams_preservesMargins() {
```

Please apply naming pattern to the the tests below.

File: app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindingAdaptersTest.kt [hunk 5]
```diff
@@ -292,6 +296,78 @@ class MarginBindingAdaptersTest {
...
+      // Set initial margins
+      setLayoutMarginStart(textView, /* marginStart= */ 24f)
+      setLayoutMarginEnd(textView, /* marginEnd= */ 40f)
+      setLayoutMarginTop(textView, /* marginTop= */ 16f)
+      setLayoutMarginBottom(textView, /* marginBottom= */ 32f)
+
+      return@runWithActivity textView
+    }
+
+    // Verify that the margins are correctly set after method calls
```
Comment (adhiamboperes) [inline app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindingAdaptersTest.kt]: Which method calls?

**Review by adhiamboperes — CHANGES_REQUESTED** [review]
Thanks @TanishMoral11! PTAL.

File: app/src/main/java/org/oppia/android/app/databinding/MarginBindingAdapters.java (hunk 1 above)
Comment (TanishMoral11) [reply app/src/main/java/org/oppia/android/app/databinding/MarginBindingAdapters.java:28]: Done.

File: app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindingAdaptersTest.kt (hunk 5 above)
Comment (TanishMoral11) [reply app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindingAdaptersTest.kt]: Thanks for the feedback, @adhiamboperes.  

The comment "// Verify that the margins are correctly set after method calls" refers to the following method calls:  
- `setLayoutMarginStart()`  
- `setLayoutMarginEnd()`  
- `setLayoutMarginTop()`  
- `setLayoutMarginBottom()`  

If you feel this comment is causing confusion, I’m happy to remove it or revise it for better clarity. 
Please let me know your preference.

File: app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindingAdaptersTest.kt (hunk 4 above)
Comment (TanishMoral11) [reply app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindingAdaptersTest.kt]: Done.

File: app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindingAdaptersTest.kt (hunk 3 above)
Comment (TanishMoral11) [reply app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindingAdaptersTest.kt]: Done.

File: app/src/main/java/org/oppia/android/app/databinding/MarginBindingAdapters.java (hunk 2 above)
Comment (TanishMoral11) [reply app/src/main/java/org/oppia/android/app/databinding/MarginBindingAdapters.java]: Thank you, @adhiamboperes , for the feedback.  

The changes occurred as a result of pressing `Ctrl + Alt + L`, which automatically formatted the code.  

No worries, I have now corrected the comments as per your guidance.

**Comment by TanishMoral11** [comment]
Hey @adhiamboperes , PTAL .

**Review by adhiamboperes — APPROVED** [review]
This LGTM, thanks @TanishMoral11!

File: app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindingAdaptersTest.kt (hunk 5 above)
Comment (adhiamboperes) [reply app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindingAdaptersTest.kt]: I think they can be safely removed.

Comment (TanishMoral11) [reply app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindingAdaptersTest.kt]: Thanks @adhiamboperes for the clarification. I’ve removed the comment to improve clarity.

**Comment by TanishMoral11** [comment]
Thanks @adhiamboperes , PTAL

**Comment by TanishMoral11** [comment]
Hey @adhiamboperes , PTAL .

**Review by adhiamboperes — APPROVED** [review]
Thanks @TanishMoral11! LGTM.

**Pull request merged**